power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -8,26 +8,120 @@
+}
+
+template <std::derived_from<Regulator> ComponentType, typename State>
+    requires main_core::component_container_c<typename State::ComponentContainer, ComponentType>
+constexpr auto get_topology_index(State const& state, auto const& id_or_index) {
+    return main_core::get_component_sequence<Regulator>(state, id_or_index);
+}
+
+template <std::derived_from<Branch> ComponentType, typename State>
+    requires main_core::component_container_c<typename State::ComponentContainer, ComponentType>
+constexpr auto get_branch_nodes(State const& state, Idx topology_sequence_idx) {
+    return state.comp_topo->branch_node_idx[topology_sequence_idx];
+}
+
+template <std::derived_from<Branch3> ComponentType, typename State>
+    requires main_core::component_container_c<typename State::ComponentContainer, ComponentType>
+constexpr auto get_branch_nodes(State const& state, Idx topology_sequence_idx) {
+    return state.comp_topo->branch3_node_idx[topology_sequence_idx];
+}
+
+template <transformer_c ComponentType, typename State>
+    requires main_core::component_container_c<typename State::ComponentContainer, ComponentType> &&
+             requires(State const& state, Idx const i) {
+                 { get_branch_nodes<ComponentType>(state, i)[i] } -> std::convertible_to<Idx>;
+             }
+inline auto get_topo_node(State const& state, Idx topology_index, ControlSide control_side) {
+    auto const& nodes = get_branch_nodes<ComponentType>(state, topology_index);
+    auto const control_side_idx = static_cast<Idx>(control_side);
+
+    assert(0 <= control_side_idx);
+    assert(control_side_idx < static_cast<Idx>(nodes.size()));
+
+    return nodes[control_side_idx];
+}
+
+template <std::derived_from<Node> ComponentType, typename State>
+    requires main_core::component_container_c<typename State::ComponentContainer, ComponentType>
+constexpr auto get_math_id(State const& state, Idx topology_sequence_idx) {
+    return state.topo_comp_coup->node[topology_sequence_idx];
+}
+
+template <std::derived_from<Branch> ComponentType, typename State>
+    requires main_core::component_container_c<typename State::ComponentContainer, ComponentType>
+constexpr auto get_math_id(State const& state, Idx topology_sequence_idx) {
+    return state.topo_comp_coup->branch[topology_sequence_idx];
+}
+
+template <std::derived_from<Branch3> ComponentType, typename State>
+    requires main_core::component_container_c<typename State::ComponentContainer, ComponentType>
+constexpr auto get_math_id(State const& state, Idx topology_sequence_idx) {
+    return state.topo_comp_coup->branch3[topology_sequence_idx];
+}
+
+template <std::derived_from<Regulator> ComponentType, typename State>
+    requires main_core::component_container_c<typename State::ComponentContainer, ComponentType>
+constexpr auto get_math_id(State const& state, Idx topology_sequence_idx) {
+    return state.topo_comp_coup->regulator[topology_sequence_idx];
+}
```

Comment:
should be in `main_core` because it involves statefullness, but i 100% agree with your comment. initially kept this here to localize the changes compared to other things that are invasive to main model.

how do we want to name this? also part of `main_core/state_queries.hpp`?